### ISSUE TITLE
add a mod selection for solo games and server admins

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#1507]
Currently, to load a mod for solo play, one have to enter console and know the magic `set extrapaks $mod1 $mod2 $mod3` incantation.
I think being able to load paks from the GUI, at least for solo games and maybe also for server admins would be good.

For a start, I would suggest a list of available mods in the `Start local/LAN game` and checkboxes near them, without handling the dependencies at all.

For dependencies, I see 2 different ways to handle them on UI, which might be complementary:

* automatically add grey "disabled" ticks in checkboxes when selecting a mod which requires another one (this seems the easier to do, while still requiring to know a mod's dependencies before loading it)
* a different list in which mods which are required by selected mods are _moved_ in

Honestly, none of those solutions are really that satisfying to me, because it's still hard to know why exactly a mod was automatically selected, but even without them, this would still be a good start and I think it would make it easier for "game admins" to switch mods (either just because or because players asked for it).

Note that I am *not* talking about a voting GUI, which would be more complex, but such UI could be used as a base for a such voting UI.

**Comment by illwieckz** [comment]
I think we may use the classic UI of two lists, one for available mods on the left, one for enabled mods on the right, with arrow buttons to move from left to right and the other way.

**Comment by Gireen** [comment]
The easiest solution is to have a list of available mods as a drop down menu together with the map selection.
That way its only possible to use a single mod at a time but anything more becomes much more complicated.

For example Mods can overwrite each other so the order they are loaded can be important and requires a more complicated UI.

**Comment by bmorel** [comment]
I don't really like dropdown boxes, as it's rather difficult to navigate them. There are not much mods yet, so it's not a big problem, but for maps it's starting to show. In my case, I have 29 custom maps and 11 official maps, for a total of 39 entries (parpax have 2 versions, older is not shown).
Browsing all the maps, selecting them to get a screenshot, which is easier to remember than the map names, is quite tedious.

This, of course, does not apply to mods, since they don't have screenshots (well, this could be imagined though, to give them a visual) and there's not many. In my case, I have the juggernaut, the betterai, the navupgrade (extends betterai), exgp (not prefixed by mod), trem (same) and finally unvanquished (same, no prefix), so that's only 6, and the betterai and  navupgrade are, certainly, hoped to be merged, but I still think dropdowns are not very practical for browsing.
Would probably be better to have a list with a single selection instead, maybe showing the visuals or description (if that's possible?) in a nearby area?

**Comment by illwieckz** [comment]
> For example Mods can overwrite each other so the order they are loaded can be important and requires a more complicated UI.

This is why I suggested the classic two list UI in my previous comment. Let's do a drawing:

```
 _____________________________________________
|                      |                      |
|    Available mods    |     Enabled mods     |
|______________________|______________________|
|                      |                      |
| exgp             [→] | [←][↑][↓] betterai   |
| jugger           [→] | [←][↑][↓] navupgrade |
| tremita          [→] |                      |
|                      |                      |
|                      |                      |
|                      |                      |
|                      |                      |
|                      |                      |
|                      |                      |
|                      |                      |
|                      |                      |
|______________________|______________________|
```

Anyway, I don't think we need to provide such complex feature. It's usually not really expected to load more than one mod. Advanced users can still do magic with `fs_extrapaks`, users would just want to load a `jugger` mod or a `tremlous` mod, and then a radio list with a cvar containing only one value would be enough:

```
Available mods:

 (*) betterai
 ( ) exgp
 ( ) betterai
 ( ) jugger
 ( ) tremita
```

The `navupgrade` mod is not really a mod, it's an extra asset package (maybe it would be better to use `res-` prefix), and I find it OK to not have UI for this kind of extra package. Or if we do an UI for `fs_extrapaks`, that would be in addition to mods.

**Comment by bmorel** [comment]
+1 for radio box list.

> The navupgrade mod is not really a mod

Indeed, and IIRC the betterai one depends on it, and as I said, it's one of thosethat I want to disappear because of a merge. Or, at least, for most of the data it contains (it's mostly to let bots crouch after all, the navcons are not in it yet, IIRC)

**Comment by slipher** [comment]
How do you propose to identify "mods"? Paks with an sgame binary in them?

There's the issue that you can't automatically know what pak dependencies are needed for a mod, because server owners might be using fs_extrapaks instead of DEPS to load the other paks. @bmorel's bot server is an extant example - it serves both `mod-betterai` and `mod-navupgrade paks`, relying on fs_extrapaks not DEPS.

**Comment by bmorel** [comment]
Yes, and I only fixed that issue on betterai mod recently. Still need to upload it, though, and I think this kind of behavior should be considered a bug in the mod.

I suppose the only reasonably easy way would be to have a naming convention for those, and to stick to it. It's a limitation, sure,  but I don't really see how to avoid this.
This also makes it easier in file system to get a look at what the files do, and to allow users to clean old versions. My betterai is again a good example here, since it arrived at 16th iteration, meaning a lot file can be removed if someone followed it since the start (that's unlikely, sure, but still).

The `map-` prefix convention is already rather widely used, same for `res-` (except that I probably should use it for navupgrade, too, and again I consider this a bug, that I'll fix soon), so I think it's not a bad idea to rely on one more prefix added to the convention.

**Comment by illwieckz** [comment]
> How do you propose to identify "mods"?

Paks with `mod-` prefix.

> There's the issue that you can't automatically know what pak dependencies are needed for a mod, because server owners might be using fs_extrapaks instead of DEPS to load the other paks. @bmorel's bot server is an extant example - it serves both `mod-betterai` and `mod-navupgrade paks`, relying on fs_extrapaks not DEPS.

It looks like there are two needs discussed there.

We need a menu to load gameplay modification as _different games_, the same way is was done in Quake 3 like it was possible to select `excessiveplus` (modification) or complete conversions (urban terror). We are not yet defining how to handle total conversions, but we can start with gameplay modifications. This may even be a specific cvar with unique name stored in it.

The thing is that `navupgrade` is something else. It can be considered as a mod, but that does not have to be the same menu.  Same would happens with texture mods. You don't chose a gameplay by selecting it.

Even if we may be able to select gameplay mods with an UI that allows to write an entire `fs_extrapaks` string, that would be over complex and then not fulfill the primary need of users wanting to change the game like when people switched from vanilla Quake 3 to Excessive Plus or Urban Terror when it was a mod.

**Comment by illwieckz** [comment]
> I suppose the only reasonably easy way would be to have a naming convention for those, and to stick to it. It's a limitation, sure, but I don't really see how to avoid this.

It's fine to enforce such naming convention, we already do it for maps. It even allows to list maps without reading dpk content. It would be the same for mods.

> The `map-` prefix convention is already rather widely used.

It's even enforced, a map is not listed if not using that prefix.

For `res-` and `tex-` those prefixes are only a convention, nothing is enforced, that's because they are dependencies of “something you start”, they are not “something you start” (like a mod, or a map).

**Comment by slipher** [comment]
Even in the case where there is only one mod pak, nother uncertainty with loading mods is whether the server host is doing it by putting the mod as the basepak or by keeping `unvanquished` as basepak and putting the mod in fs_extrapaks.

**Comment by bmorel** [comment]
There is a `basepak`? Should not mods be basepaks which depends on the game directly, then, instead of using extrapaks?
Then, extrapaks would, in current status, mostly be about changing textures and configurations, but clearly nowadays, all the "serious" mods seem to imply code changes, and thus should be basepaks depending on unvanquished?
We have, I think, a real documentation issue on that... (if someone explains, I'm okay to write some wiki on the topic though)

If I am right about the `basepack` (that is, it's the root of the graph, and depends on all resources it requires) then it's even simpler. Only allow to decide which `basepack` is to use, and the modders will have responsibility to properly define the DEPS file.
The tweaking of resources, textures, models, etc, though UI can come later, for now we just don't have any other way to implement different winning conditions than a "total conversion" anyway.

**Comment by bmorel** [comment]
there is currently a way for solo players to write the name of a mod when starting a new local game. It's very limited (no auto discovery like a dropbox or whatever would provide) though.
